File: src/edu/stanford/cfuller/imageanalysistools/image/ImageCoordinate.java
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is Colin J. Fuller's code.
 *
 * The Initial Developer of the Original Code is
 * Colin J. Fuller.
 * Portions created by the Initial Developer are Copyright (C) 2011
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s): Colin J. Fuller
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

package edu.stanford.cfuller.imageanalysistools.image;

import java.util.Collection;
import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Represents an integer-valued coordinate in a 5D image (X, Y, Z, color, time).
 *
 * <p>
 * To avoid unnecessary object allocation when iterating over large images, this class provides basic object pooling capabilities.
 * ImageCoordinates can only be created using this class's static factory methods ({@link #createCoord}, {@link #cloneCoord}), which will reuse ImageCoordinates if there are any
 * available or allocate more as necessary.
 *
 * <p> Users of this class are responsible for calling {@link #recycle} on any ImageCoordinate they explicitly create using one of the static factory methods
 * but not any coordinates that they obtain by any other means.  Calling recycle on an ImageCoordinate will return it to the pool of available coordinates.
 * ImageCoordinates are not guaranteed to retain their value after being recycled, so they should not be used after being recycled.
 *
 * <p> This class is partly thread-safe in the sense that multiple threads can use the same pool of ImageCoordinates safely.  However, ImageCoordinate objects themselves
 * are not thread-safe and should not be shared between threads.
 *
 *
 * @author Colin J. Fuller
 *
 */

public class ImageCoordinate implements java.io.Serializable, Collection<Integer> {

	static final long serialVersionUID = 1L;
	
	static final int initialStaticCoordCount = 8;
	
	private static java.util.ArrayList<ImageCoordinate> availableStaticCoords;
	private static java.util.Hashtable<ImageCoordinate, Boolean> inUse;
	
	private int x;
	private int y;
	private int z;
	private int c;
	private int t;
	
	private final int dimensionality = 5;
			
	static {

        synchronized(ImageCoordinate.class) {

            availableStaticCoords = new java.util.ArrayList<ImageCoordinate>();
            inUse = new java.util.Hashtable<ImageCoordinate, Boolean>();

            for (int i = 0; i < initialStaticCoordCount; i++) {
                ImageCoordinate c = new ImageCoordinate();
                inUse.put(c, false);
                availableStaticCoords.add(c);

            }

        }
	}
	
	private ImageCoordinate(){}

    /**
     * Gets the x-component of this ImageCoordinate.
     * @return  The x-component as an integer.
     */
	public int getX(){return this.x;}

    /**
     * Gets the y-component of this ImageCoordinate.
     * @return  The y-component as an integer.
     */
	public int getY(){return this.y;}

    /**
     * Gets the z-component of this ImageCoordinate.
     * @return  The z-component as an integer.
     */
	public int getZ(){return this.z;}

    /**
     * Gets the c-component of this ImageCoordinate.
     * @return  The c-component as an integer.
     */
	public int getC(){return this.c;}

    /**
     * Gets the t-component of this ImageCoordinate.
     * @return  The t-component as an integer.
     */
	public int getT(){return this.t;}

    /**
     * Factory method that creates an ImageCoordinate with the specified coordinates.
     *
     * This method will attempt to recycle an existing ImageCoordinate if any are available, or create a new one if none are available.
     *
     * Users of this method should call {@link #recycle} on the ImageCoordinate returned by this method when finished with it in order to
     * make the coordinate available for reuse on other calls of this method.
     *
     * @param x     The x-component of the ImageCoordinate to be created.
     * @param y     The y-component of the ImageCoordinate to be created.
     * @param z     The z-component of the ImageCoordinate to be created.
     * @param c     The c-component of the ImageCoordinate to be created.
     * @param t     The t-component of the ImageCoordinate to be created.
     * @return      An ImageCoordinate whose components have been set to the specified values.
     */
	public static synchronized ImageCoordinate createCoord(int x, int y, int z, int c, int t) {
		
		ImageCoordinate staticCoord = null;

		if (! availableStaticCoords.isEmpty()) {
			
			staticCoord = availableStaticCoords.remove(0);
			
//			if (inUse.get(staticCoord)) {
//				Logger.getLogger("edu.stanford.cfuller.imageanalysistools").warning("WARNING! Image coordinate was likely recycled twice: " + staticCoord.toString() + "  This can only result from a coding bug.");
//				Logger.getLogger("edu.stanford.cfuller.imageanalysistools").warning("at:");
//                try {
//                    throw new IllegalStateException();
//                } catch (IllegalStateException e) {
//                    e.printStackTrace();
//                }
//                staticCoord = new ImageCoordinate();
//			} 

		} else {
			staticCoord = new ImageCoordinate();
		}
		
		staticCoord.x = x;
		staticCoord.y = y;
		staticCoord.z = z;
		staticCoord.c = c;
		staticCoord.t = t;
//		inUse.put(staticCoord, true);
		return staticCoord;
		
	}


    /**
     * Factory method that creates and ImageCoordinate that represents the same position in coordinate space as the supplied coordinate.
     *
     * Users of this method should call {@link #recycle} on the ImageCoordinate returned by this method when finished with it in order to
     * make the coordinate available for reuse on other calls of this method.
     *
     * @param toClone   The ImageCoordinate to clone.
     * @return          An ImageCoordinate whose components are set to the same values as the components of the supplied ImageCoordinate.
     */
    public static ImageCoordinate cloneCoord(ImageCoordinate toClone) {

        return createCoord(toClone.getX(), toClone.getY(), toClone.getZ(), toClone.getC(), toClone.getT());

    }

    /**
     * Returns an ImageCoordinate to the pool of ImageCoordinates available for reuse.
     *<p>
     * This method should be called on an ImageCoordinate if and only if that coordinate was obtained using one of the static factory methods
     * ({@link #createCoord} or {@link #cloneCoord}).
     *<p>
     * Each ImageCoordinate must only be recycled once, or else future calls to the factory methods may return the same coordinate multiple times with
     * different values, which will lead to unpredictable behavior.  No checking is done to ensure that each coordinate was only recycled once (this was found
     * to cause enough overhead that the benefits of reusing objects were cancelled out).
     *
     *
     */
	public void recycle() {

        synchronized(ImageCoordinate.class) {

            /*if (availableStaticCoords.contains(this)) {
                Logger.getLogger("edu.stanford.cfuller.imageanalysistools").warning("WARNING! Possible second recycle of: " + this.toString() + "  This can only result from a coding bug.");
                Logger.getLogger("edu.stanford.cfuller.imageanalysistools").warning("at:");
                try {
                    throw new IllegalStateException();
                } catch (IllegalStateException e) {
                    e.printStackTrace();
                }
            } else {
            */
            availableStaticCoords.add(this);
            //}
//          inUse.put(this, false);
        }
	}

    /**
     * Converts the ImageCoordinate to a string representation.
     * <p>
     * The returned string will be "(x,y,z,c,t) = (a,b,c,d,e)", where a through e are placeholders for the actual coordinates of the ImageCoordinate.
     * @return  The string representation of the ImageCoordinate.
     */
	public String toString() {
		String result = "(x,y,z,c,t) = (" + Integer.toString(this.x)+ ", "+ Integer.toString(this.y) + ", " + Integer.toString(this.z)+ ", " + Integer.toString(this.c) + ", " + Integer.toString(this.t) + ")";
		return result;
	}

    /**
     * Sets all the components of the ImageCoordinate to the specified values.
     * 
     * @param x     The new x-component of the ImageCoordinate.
     * @param y     The new y-component of the ImageCoordinate.
     * @param z     The new z-component of the ImageCoordinate.
     * @param c     The new c-component of the ImageCoordinate.
     * @param t     The new t-component of the ImageCoordinate.
     */
	public void setCoord(int x, int y, int z, int c, int t) {
		this.x = x;
		this.y = y;
		this.z = z;
		this.c = c;
		this.t = t;
	}

    /**
     * Sets the components of the ImageCoordinate to the values of the components of another ImageCoordinate.
     * @param other     The ImageCoordinate whose component values will be copied.
     */
    public void setCoord(ImageCoordinate other) {
        this.setCoord(other.getX(), other.getY(), other.getZ(), other.getC(), other.getT());
    }

    /**
     * Sets the x-component of the ImageCoordinate to the specified value.
     * @param x The new x-component of the ImageCoordinate.
     */
	public void setX(int x) {
		this.x = x;
	}

    /**
     * Sets the y-component of the ImageCoordinate to the specified value.
     * @param y The new y-component of the ImageCoordinate.
     */
	public void setY(int y) {
		this.y = y;
	}

    /**
     * Sets the z-component of the ImageCoordinate to the specified value.
     * @param z The new z-component of the ImageCoordinate.
     */
	public void setZ(int z) {
		this.z = z;
	}

    /**
     * Sets the c-component of the ImageCoordinate to the specified value.
     * @param c The new c-component of the ImageCoordinate.
     */
	public void setC(int c) {
		this.c = c;
	}

    /**
     * Sets the t-component of the ImageCoordinate to the specified value.
     * @param t The new t-component of the ImageCoordinate.
     */
	public void setT(int t) {
		this.t = t;
	}
	
	/**
	 * Gets the coordinate specifies by the supplied index, ordered in the same
	 * way as when iterating over the coordinates.
	 * @param index		The index of the coordinate to get.
	 * @return			The coordinate at the specified index.
	 * @throws ArrayIndexOutOfBoundsException if the index specified is not >= 0 and less than the dimensionality of the ImageCoordiante.
	 */
	public Integer get(int index) {
		switch (index) {
		case 0:	return getX();
		case 1: return getY();
		case 2: return getZ();
		case 3: return getC();
		case 4: return getT();
		default: break;
		}
		throw new ArrayIndexOutOfBoundsException("Invalid coordinate index.  Dimensionality: " + dimensionality);
	}
	
	
	/**
	 * Sets the coordinate specifies by the supplied index, ordered in the same
	 * way as when iterating over the coordinates.
	 * @param index	The index of the coordinate to get.
	 * @param value	The value to which the specified coordinate will be set.
	 * @throws ArrayIndexOutOfBoundsException if the index specified is not >= 0 and less than the dimensionality of the ImageCoordiante.
	 */
	public void set(int index, int value) {
		switch (index) {
		case 0:	setX(value); return;
		case 1: setY(value); return;
		case 2: setZ(value); return;
		case 3: setC(value); return;
		case 4: setT(value); return;
		default: break;
		}
		throw new ArrayIndexOutOfBoundsException("Invalid coordinate index.  Dimensionality: " + dimensionality + ".  Index: " + index);
	}
	
	/**
	 * Gets the dimensionality of the ImageCoordinate (that is, the number of dimensions with coordinates stored).
	 * @return	The dimensionality, as an int.
	 */
	public int getDimensionality() {
		return this.dimensionality;
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#add(java.lang.Object)
	 */
	@Override
	public boolean add(Integer arg0) {
		throw new UnsupportedOperationException("Add not supported for ImageCoordinates.");
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#addAll(java.util.Collection)
	 */
	@Override
	public boolean addAll(Collection<? extends Integer> arg0) {
		throw new UnsupportedOperationException("Add not supported for ImageCoordinates.");
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#clear()
	 */
	@Override
	public void clear() {
		throw new UnsupportedOperationException("Add not supported for ImageCoordinates.");
		
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#contains(java.lang.Object)
	 */
	@Override
	public boolean contains(Object arg0) {
		Integer iArg = (Integer) arg0;
		for (Integer i : this) {
			if (i == iArg) return true;
		}
		return false;
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#containsAll(java.util.Collection)
	 */
	@Override
	public boolean containsAll(Collection<?> arg0) {
		for (Object o : arg0) {
			if (!this.contains(o)) return false;
		}
		return true;
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#isEmpty()
	 */
	@Override
	public boolean isEmpty() {
		return false;
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#iterator()
	 */
	@Override
	public Iterator<Integer> iterator() {
		return new ImageCoordinateIterator();
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#remove(java.lang.Object)
	 */
	@Override
	public boolean remove(Object arg0) {
		throw new UnsupportedOperationException("Remove not supported for ImageCoordinates.");

	}

	/* (non-Javadoc)
	 * @see java.util.Collection#removeAll(java.util.Collection)
	 */
	@Override
	public boolean removeAll(Collection<?> arg0) {
		throw new UnsupportedOperationException("Remove not supported for ImageCoordinates.");

	}

	/* (non-Javadoc)
	 * @see java.util.Collection#retainAll(java.util.Collection)
	 */
	@Override
	public boolean retainAll(Collection<?> arg0) {
		throw new UnsupportedOperationException("Retain not supported for ImageCoordinates.");

	}

	/* (non-Javadoc)
	 * @see java.util.Collection#size()
	 */
	@Override
	public int size() {
		return dimensionality;
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#toArray()
	 */
	@Override
	public Object[] toArray() {
		return toArray(new Object[0]);
	}

	/* (non-Javadoc)
	 * @see java.util.Collection#toArray(T[])
	 */
	@SuppressWarnings("unchecked")
	@Override
	public <T> T[] toArray(T[] arg0) {
		if (arg0.length < this.size()) {
			int c = 0; 
			for (Integer i : this) {
				arg0[c++] = (T) i;
			}
			return arg0;
		} else {
			Integer[] toReturn = new Integer[this.size()];
			int c = 0;
			for (Integer i : this) {
				toReturn[c++] = i;
			}
			return (T[]) toReturn;
		}
	}
	
	protected class ImageCoordinateIterator implements Iterator<Integer> {
		int currentIndex;
		
		public ImageCoordinateIterator() {
			this.currentIndex = 0;
		}
		
		public boolean hasNext() {
			if (this.currentIndex < dimensionality) {
				return true;
			}
			return false;
		}
		
		public Integer next() {
			switch (currentIndex++) {
			case 0:	return getX();
			case 1: return getY();
			case 2: return getZ();
			case 3: return getC();
			case 4: return getT();
			default: break;
			}
			
			throw new NoSuchElementException("No more elements in ImageCoordinate.");
			
		}
		
		public void remove() {
			throw new UnsupportedOperationException("Remove not supported for ImageCoordinate.");
		}
		
		
	}
	
}

File: src/main/java/edu/stanford/cfuller/imageanalysistools/image/ImageCoordinate.kt
package edu.stanford.cfuller.imageanalysistools.image

import java.util.NoSuchElementException

/**
 * Represents an integer-valued coordinate in an n-dimensional image.
 *
 * To avoid unnecessary object allocation when iterating over large images, this class provides basic object pooling capabilities.
 * ImageCoordinates can only be created using this class's static factory methods ([.createCoord], [.createCoordXYZCT], [.cloneCoord]), which will reuse ImageCoordinates if there are any
 * available or allocate more as necessary.
 *
 * Users of this class are responsible for calling [.recycle] on any ImageCoordinate they explicitly create using one of the static factory methods
 * but not any coordinates that they obtain by any other means.  Calling recycle on an ImageCoordinate will return it to the pool of available coordinates.
 * ImageCoordinates are not guaranteed to retain their value after being recycled, so they should not be used after being recycled.
 *
 * This class is partly thread-safe in the sense that multiple threads can use the same pool of ImageCoordinates safely.  However, ImageCoordinate objects themselves
 * are not thread-safe and should not be shared between threads.
 *
 * Arbitrary dimensions are defined by an int identifier, which can be any valid int.  This class defines constants for X,Y,Z,C,T as 0 to 4.
 * When defining additional dimensions, using lower numbers and not leaving gaps will improve performance.
 *
 * @author Colin J. Fuller
 */

class ImageCoordinate private constructor() : java.io.Serializable, Collection<Int> {

    private var dimensionCoordinates: IntArray = intArrayOf()
    private var undefinedCoordinates: ByteArray = byteArrayOf()
    /**
     * Gets the dimension of the ImageCoordinate: that is, the number of defined named dimensions
     * it contains.
     * @return        the dimension of the ImageCoordinate.
     */
    var dimension: Int = 0
        private set

    init {
        this.dimensionCoordinates = IntArray(initialDimensionCapacity)
        this.undefinedCoordinates = ByteArray(this.dimensionCoordinates.size)
        java.util.Arrays.fill(this.undefinedCoordinates, ImageCoordinate.one)
        this.dimension = 0

    }

    /**
     * Gets the specified coordinate by index.
     *
     * The order of coordinates will be the order in which they were set if done manually,
     * or the order from the coordinate this ImageCoordinate was cloned from,
     * or otherwise ordered unpredictably (though consistent between calls), probably
     * according to hash key.
     *
     * If the coordinate index is out of bounds, returns 0 instead of throwing an exception.
     * @param dimensionConstant    the constant corresponding to the dimension whose value will be retrieved.
     * @return                    the value of the specified dimension component, or 0 if the index was out of bounds.
     */
    operator fun get(dimensionConstant: Int): Int {
        if (dimensionConstant < this.dimensionCoordinates.size) {
            return this.dimensionCoordinates[dimensionConstant]
        }
        return 0
    }

    /**
     * Sets the specified coordinate component of the ImageCoordinate by its index.
     *
     * This checks that the dimension exists and adds it if it does not exist.
     * @param dimensionConstant        the constant corresponding to the dimension to set.
     * @param value                    the value to which to set the coordinate.
     */
    operator fun set(dimensionConstant: Int, value: Int) {
        if (dimensionConstant >= this.dimensionCoordinates.size) {
            val oldCoords = this.dimensionCoordinates
            val oldUndefined = this.undefinedCoordinates
            val newSize = if (dimensionConstant + 1 > 2 * this.dimensionCoordinates.size) {
                dimensionConstant + 1
            } else { 2 * this.dimensionCoordinates.size }
            this.dimensionCoordinates = IntArray(newSize)
            this.undefinedCoordinates = ByteArray(newSize)

            java.util.Arrays.fill(this.dimensionCoordinates, 0)
            java.util.Arrays.fill(this.undefinedCoordinates, ImageCoordinate.one)

            for (i in oldCoords.indices) {
                this.dimensionCoordinates[i] = oldCoords[i]
                this.undefinedCoordinates[i] = oldUndefined[i]
            }
        }
        this.dimension += this.undefinedCoordinates[dimensionConstant].toInt()
        this.undefinedCoordinates[dimensionConstant] = ImageCoordinate.zero
        this.dimensionCoordinates[dimensionConstant] = value
    }

    /**
     * Sets the specified coordinate component of the ImageCoordinate by its index.
     *
     * This does not check that the dimension exists before accessing it.
     * @param dimensionConstant        the constant corresponding to the dimension to set.
     * @param value                    the value to which to set the coordinate.
     */
    fun quickSet(dimensionConstant: Int, value: Int) {
        this.dimensionCoordinates[dimensionConstant] = value
    }

    /**
     * Gets the specified coordinate component of the ImageCoordinate by its index.
     *
     * This does not check that the dimension exists before accessing it.
     * @param dimensionConstant        the constant corresponding to the dimension whose value will be retrieved.
     * @return                        the value of the specified dimension component
     */
    fun quickGet(dimensionConstant: Int): Int {
        return this.dimensionCoordinates[dimensionConstant]
    }

    /**
     * Returns an ImageCoordinate to the pool of ImageCoordinates available for reuse.
     *
     * This method should be called on an ImageCoordinate if and only if that coordinate was obtained using one of the static factory methods
     * ([.createCoord] or [.cloneCoord]).
     *
     * Each ImageCoordinate must only be recycled once, or else future calls to the factory methods may return the same coordinate multiple times with
     * different values, which will lead to unpredictable behavior.  No checking is done to ensure that each coordinate was only recycled once (this was found
     * to cause enough overhead that the benefits of reusing objects were cancelled out).
     */
    fun recycle() {
        synchronized(ImageCoordinate::class.java) {
            availableStaticCoords!!.add(this)
        }
    }

    /**
     * Gets the defined index of the specified coordinate.
     * @param i        The coordinate to retrieve (i.e. the value i such that the ith index that has been assigned a value will be returned.)
     * @return        The index of the ith component.
     */
    fun getDefinedIndex(i: Int): Int? {
        var c = -1
        for (j in this.undefinedCoordinates.indices) {
            c += 1 - this.undefinedCoordinates[j]
            if (c == i) return j
        }
        throw NoSuchElementException("The specified coordinate is not defined: " + i + ".  Number of defined coordinates: " + this.dimension)
    }

    /**
     * Converts the ImageCoordinate to a string representation.
     *
     * The returned string will be "(a,b,c,d,e)", where a through e
     * are placeholders for the actual coordinates of the ImageCoordinate.
     * @return  The string representation of the ImageCoordinate.
     */
    override fun toString(): String {
        if (this.dimension == 0) {
            return "()"
        }
        var valueString = "("
        for (i in this) {
            valueString += this[i].toString() + ", "
        }
        valueString = valueString.substring(0, valueString.length - 2) // strip off the last ", "
        valueString += ")"
        return valueString
    }

    /**
     * Sets all the components of the ImageCoordinate to the specified values.
     * @param x     The new x-component of the ImageCoordinate.
     * @param y     The new y-component of the ImageCoordinate.
     * @param z     The new z-component of the ImageCoordinate.
     * @param c     The new c-component of the ImageCoordinate.
     * @param t     The new t-component of the ImageCoordinate.
     */
    fun setCoordXYZCT(x: Int, y: Int, z: Int, c: Int, t: Int) {
        this[ImageCoordinate.X] = x
        this[ImageCoordinate.Y] = y
        this[ImageCoordinate.Z] = z
        this[ImageCoordinate.C] = c
        this[ImageCoordinate.T] = t
    }

    /**
     * Sets the components of the ImageCoordinate to the values of the components of another ImageCoordinate.
     * @param other     The ImageCoordinate whose component values will be copied.
     */
    fun setCoord(other: ImageCoordinate) {
        this.clear()
        for (i in other) {
            this[i] = other[i]
        }
    }

    /**
     * Clears all stored coordinate dimensions and values.
     *
     * After calling this method, the ImageCoordinate will be zero-dimensional.
     */
    fun clear() {
        java.util.Arrays.fill(this.undefinedCoordinates, ImageCoordinate.one)
        this.dimension = 0
    }

    override operator fun contains(element: Int): Boolean {
        return this.any { it == element }
    }

    override fun containsAll(elements: Collection<Int>): Boolean {
        return elements.all { this.contains(it) }
    }

    override fun isEmpty(): Boolean {
        return false
    }

    override fun iterator(): Iterator<Int> {
        return ImageCoordinateIterator(this)
    }

    override val size: Int
        get() {
            return this.dimension
        }

    private class ImageCoordinateIterator(internal var ic: ImageCoordinate) : Iterator<Int> {
        internal var currentIndex: Int = 0
        internal var currentDefinedCount: Int = 0

        init {
            this.currentIndex = 0
            this.currentDefinedCount = 0
        }

        override fun hasNext(): Boolean {
            return this.currentDefinedCount < this.ic.dimension
        }

        override fun next(): Int {
            while (this.hasNext()) {
                if (ic.undefinedCoordinates[this.currentIndex++].toInt() == 0) {
                    this.currentDefinedCount += 1
                    return this.currentIndex - 1
                }
            }
            throw NoSuchElementException("No more elements in ImageCoordinate.")
        }
    }

    companion object {
        //TODO: consider whether coordinates should continue to be discrete as they are in the current implementation
        val X = 0
        val Y = 1
        val Z = 2
        val C = 3
        val T = 4
        val defaultDimensionOrder = "XYZCT"
        internal const val serialVersionUID = 1L
        internal val initialStaticCoordCount = 8
        internal val initialDimensionCapacity = 10
        private var availableStaticCoords: java.util.Deque<ImageCoordinate>? = null
        private val one: Byte = 1
        private val zero: Byte = 0

        // TODO(colin): reevaluate whether pooling is still a performance win on modern JVMs
        init {
            synchronized(ImageCoordinate::class.java) {
                availableStaticCoords = java.util.LinkedList<ImageCoordinate>()
                for (i in 0..initialStaticCoordCount - 1) {
                    val c = ImageCoordinate()
                    availableStaticCoords!!.add(c)
                }
            }
        }

        private val nextAvailableCoordinate: ImageCoordinate
            get() {
                var coord: ImageCoordinate = ImageCoordinate()
                synchronized(ImageCoordinate::class.java) {
                    if (!(availableStaticCoords?.isEmpty() ?: true)) {
                        coord = availableStaticCoords?.removeFirst() ?: coord
                    }
                }
                coord.clear()
                return coord
            }

        /**
         * Factory method that creates a 5D ImageCoordinate with the specified coordinates.
         *
         * This method will attempt to recycle an existing ImageCoordinate if any are available, or create a new one if none are available.
         * Users of this method should call [.recycle] on the ImageCoordinate returned by this method when finished with it in order to
         * make the coordinate available for reuse on other calls of this method.
         * @param x     The x-component of the ImageCoordinate to be created.
         * @param y     The y-component of the ImageCoordinate to be created.
         * @param z     The z-component of the ImageCoordinate to be created.
         * @param c     The c-component of the ImageCoordinate to be created.
         * @param t     The t-component of the ImageCoordinate to be created.
         * @return      An ImageCoordinate whose components have been set to the specified values.
         */
        fun createCoordXYZCT(x: Int, y: Int, z: Int, c: Int, t: Int): ImageCoordinate {
            val staticCoord = ImageCoordinate.nextAvailableCoordinate
            //setting T first avoids multiple allocations of internal storage for new coordinates
            staticCoord[ImageCoordinate.T] = t
            staticCoord[ImageCoordinate.X] = x
            staticCoord[ImageCoordinate.Y] = y
            staticCoord[ImageCoordinate.Z] = z
            staticCoord[ImageCoordinate.C] = c
            return staticCoord
        }

        /**
         * Factory method that creates a zero-dimensional ImageCoordinate that can be manually extended.
         *
         * This method will attempt to recycle an existing ImageCoordinate if any are available, or create a new one if none are available.
         * Users of this method should call [.recycle] on the ImageCoordinate returned by this method when finished with it in order to
         * make the coordinate available for reuse on other calls of this method.
         * @return      An ImageCoordinate with no specified dimensions.
         */
        fun createCoord(): ImageCoordinate {
            return ImageCoordinate.nextAvailableCoordinate
        }

        /**
         * Factory method that creates and ImageCoordinate that represents the same position in coordinate space as the supplied coordinate.
         * Users of this method should call [.recycle] on the ImageCoordinate returned by this method when finished with it in order to
         * make the coordinate available for reuse on other calls of this method.
         * @param toClone   The ImageCoordinate to clone.
         * @return          An ImageCoordinate whose components are set to the same values as the components of the supplied ImageCoordinate.
         */
        fun cloneCoord(toClone: ImageCoordinate): ImageCoordinate {
            val newCoord = ImageCoordinate.nextAvailableCoordinate
            newCoord.setCoord(toClone)
            return newCoord
        }
    }
}
